chatmeleon-server/src/controllers/message-controller.ts

```diff
@@ -1,114 +1,252 @@
+  const pageSize = parseInt(req.query.pageSize as string, 10) || 10; // Get desired page size from query parameters
+  const cursor = req.query.cursor as string | undefined; // Get optional cursor for pagination from query parameters
 
-  const allMessages = await prisma.message.findMany({
+  const messages = await prisma.message.findMany({
     where: {
       sender: {
-        id: userId,
+        id: userId, // Filter messages sent by the authenticated user
       },
       id: {
-        gt: cursor,
+        gt: cursor, // Retrieve messages after the provided cursor (if any)
       },
     },
-    take: pageSize,
+    orderBy: {
+      createdAt: "asc",
+    },
+    take: pageSize, // Limit results to the specified page size
   });
 
-  const lastMessage = allMessages[allMessages.length - 1];
+  // Extract the last message's cursor for pagination.
+  const lastMessage = messages[messages.length - 1];
   const nextCursor = lastMessage ? lastMessage.id : null;
 
   res.json({
-    allMessages,
-    nextCursor,
+    messages, // Send the fetched messages in the response
+    nextCursor, // Include the cursor for further pagination
   });
 };
 
+// Function to create a new message
 const createMessage = async (req: Request, res: Response) => {
+  // Extract data from request body and authentication
   const { body, image, conversationId } = req.body;
   const { userId: senderId } = req.auth!;
 
-  if (!(await hasConversation(senderId, conversationId))) {
-    return res
-      .status(404)
-      .type("text/plain")
-      .send("Not Found - Conversation not found.");
-  }
-
-  await prisma.message.create({
-    data: {
-      body,
-      image,
-      conversation: {
-        connect: {
-          id: conversationId,
-        },
-      },
-      sender: {
-        connect: {
-          id: senderId,
-        },
-      },
-    },
-  });
+  prisma
+    .$transaction(async (tx) => {
+      try {
+        // Validate conversation existence
+        await tx.conversation.findFirstOrThrow({
+          where: {
+            id: conversationId,
+            userIds: { has: senderId },
+          },
+        });
+
+        // Create the new message within the transaction
+        const newMessage = await tx.message.create({
+          data: {
+            body,
+            image,
+            conversation: {
+              connect: {
+                id: conversationId,
+              },
+            },
+            sender: {
+              connect: {
+                id: senderId,
+              },
+            },
+          },
+        });
+
+        // Update conversation with new message ID and timestamp
+        await tx.conversation.update({
+          where: {
+            id: conversationId,
+          },
+          data: {
+            lastMessageAt: newMessage.createdAt,
+            lastMessage: newMessage.body,
+            messagesIds: { push: newMessage.id },
+          },
+        });
```

Comment:
This works but I think all of this can be combined into 1 query but the conversation schema has to be changed. So this is just a comment.

Update throws if the where clause cannot be satisfied (` where: {
            id: conversationId,
            userIds: { has: senderId },
          },`) https://www.prisma.io/docs/orm/reference/prisma-client-reference#update

To combine the create message with the update conversation, we would need to change the conversation schema to only contain the field messageIds and we can get info on the last message from that but that has performance implications so I guess its up to the system design.